### ISSUE TITLE
Please ship App::cpanminus::script and friends in cpan installs

### ISSUE THREAD

**Issue by kentfredric** [miyagawa/cpanminus#431]
My apologies if this has been discussed before but I couldn't find a related issue.

I do consider the way `cpanm` currently does what it does has to be done for various reasons.

However, that shipping the component modules allows people who's reasons are outside those reasons to do what they want to do using the `cpanm` base code without resorting to nasty tricks.

For instance, its obvious `cpanm` doesn't want to use system `@INC`  for deps, but as a user, I'm inclined to want to turn that off from time to time, just to see what `cpanm` does using system `@INC`

And for instance, if I want to wrap the cpanm code, extend it, or apply roles to it, that is made easier by the modules in question being independent.

I'm not exactly asking for `cpanm` itself to use these shipped modules, and of course, there's scope for confusion if somebody modifies shipped versions of `App::cpanminus::script` under the illusion that will have an effect on the shipped `cpanm` client, so it will need some obligatory `hic sunt dracones` warnings.

But that option would be much preferable to the sorts of hacks we're considering to simply hook various `cpanm` phases for custom behaviour.

For instance, this sort of stuff is pretty horrible and inflexible: 
- https://metacpan.org/source/KPEE/App-cpanmw-0.03/script/cpanmw

And there's so many ways of making it work but none of them are what I'd call "nice"
- https://gist.github.com/kentfredric/ce1df3e7e509e071b63d
- https://gist.github.com/dolmen/3b2a03fdffd113fbf73b
- https://gist.github.com/kentfredric/0f418bb70e09abad3046

All of these examples could then be nuked from orbit and much simpler code would emerge:

```perl
package Foo;
use Moo;
extends  'App::cpanminus::script';
# The worlds your onion
```

In short, it seems viable to have the bulk of cpanm code shipped in such a way others can consume it without having to themselves disassemble the parts out of it they need, without having it directly conflict with the dominant usecases of cpanm. 

NB: It would of course be the burden of the consumer to make sure all the necessary dependencies to do this were done on their code, not cpanms.

**Comment by miyagawa** [comment]
Obviously, making them available was my initial thought to provide a programming interface for cpanm but it didn't go well due to many reasons, one of which is the `@INC` management as you mentioned.

I understand your frustration on this and the extensibility of cpanm has been pretty limited because of this design. That was one reason I implemented plugin system back in 0.99 era, but this was disabled as well since some of the users were against the idea of even extending it :/

Now,  I see two reasons to want to extend cpanm as you mentioned: 1) to be able to inspect what is going on (debugging) and 2) be able to change/extend it by third party modules.

or 1), I get the frustration, but luckily there's an easy workaround, that is to run it from git clone, via: `perl script/cpanm.PL` - it will not involve any fatpacking at all and run it with the system `@INC`.

2) is the harder problem - well, shipping it is the easiest part, but keeping the interface and internal structure is the much harder part. As you see the `App::cpanminus::script` is the God class that does everything, and its interface and data structure are not designed with the API consumption in mind. It's easy to just include it to cpanm today, but keeping the interface so that third party extensions (cpanm-reporter comes to mind) relying on that would be a much bigger promise I'd implicitly make by doing so.

My long-term goal for this is to split out all the guts of cpanminus into modules, including spitting out relevant chunks into reusable parts. CPAN::Common::Index and Dist::Metadata are close to what those reusable components would look like, although they aren't exactly split out of cpanm.

I had an earlier experiment for that somewhere on github couple of years ago, but it stalled due to other issues. Now that I started hacking on [Carmel](https://github.com/miyagawa/Carmel) and Carton which would require such functionalities more than anything (right now they use some experimental cpanm flags to get things going), I'd hopefully get it done sometime soon.

I'd keep this bug open for now since I think the concern is legit, but shipping it would create another issue that I cannot commit right now.

**Comment by miyagawa** [comment]
This is the main goal of the Menlo dist I'm working on, and will be fixed once it is shipped.

**Comment by miyagawa** [comment]
Read more about Menlo here http://weblog.bulknews.net/post/117034550339/cpanm-towards-2-0 and I'll make more documentation in the dist itself and blog more when it's actually released. There's no ETA and I suspect it's a work worth months.
